### ISSUE TITLE
Synchronous requests

### ISSUE THREAD

**Issue by gotev** [Microsoft/thrifty#145]
It will be useful to be able to make also synchronous Thrift requests, because they are useful inside Android intent services. Currently to make a call to `myFunction`, the only way is:
```java
client.myFunction(request, new ServiceMethodCallback<MyResponse>() {
    @Override
    public void onSuccess(MyResponse result) {
        Log.i("Response", result.accessToken);
    }

    @Override
    public void onError(Throwable error) {
        Log.e("Error", "Error", error);
    }
});
```
Which is async and not suitable to be used inside a service as is.

By [wrapping the callback inside a Future](https://stackoverflow.com/a/2180534), I made a workaround to make synchronous calls, although the resulting code is not as easy to read and generates some overhead as well.
```java
try {
    SyncServiceMethodCallback<MyResponse> responseFuture = new SyncServiceMethodCallback<>();

    client.myFunction(request, responseFuture);

    MyResponse result = responseFuture.get();

    Log.i("Response", result.accessToken);

} catch (Throwable error) {
    Log.e("Error", "Error", error);
}
```
The ideal will be to have:
```java
try {
    MyResponse result = client.myFunction(request);

    Log.i("Response", result.accessToken);

} catch (Throwable error) {
    Log.e("Error", "Error", error);
}
```
I would like to contribute to this great project, but I wanted to know which is the best way to make such implementation in your opinion, based on the current library design. Something like Retrofit's CallAdapterFactory will be great, so other adapters could be easily developed as plugins (e.g. Rx).

Btw, here there's `SyncServiceMethodCallback` code, just in case someone else needs it:
```java
import android.support.annotation.NonNull;
import com.microsoft.thrifty.service.ServiceMethodCallback;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class SyncServiceMethodCallback<T> implements ServiceMethodCallback<T>, Future<T> {

    private volatile T callbackResult = null;
    private volatile boolean cancelled = false;
    private volatile Throwable exception = null;
    private final CountDownLatch countDownLatch;

    public SyncServiceMethodCallback() {
        countDownLatch = new CountDownLatch(1);
    }

    @Override
    public void onSuccess(T result) {
        callbackResult = result;
        countDownLatch.countDown();
    }

    @Override
    public void onError(Throwable error) {
        exception = error;
        countDownLatch.countDown();
    }

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        if (isDone()) {
            return false;
        } else {
            countDownLatch.countDown();
            cancelled = true;
            return !isDone();
        }
    }

    @Override
    public boolean isCancelled() {
        return cancelled;
    }

    @Override
    public boolean isDone() {
        return countDownLatch.getCount() == 0;
    }

    @Override
    public T get() throws InterruptedException, ExecutionException {
        countDownLatch.await();
        if (exception != null)
            throw new ExecutionException(exception);
        return callbackResult;
    }

    @Override
    public T get(long timeout, @NonNull TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        countDownLatch.await(timeout, unit);
        if (exception != null)
            throw new ExecutionException(exception);
        return callbackResult;
    }
}
```

**Comment by gabrielittner** [comment]
I've written a thrifty compiler plugin some time ago that coverts the generated code to make synchronous calls. It still used a `Result` class which either has the response or the exception, but that's already nicer than using a Future. 

Imho the main part on thrifty's side to support synchronous requests or to at least make the post processing approach nicer would be to make this part of 
[invokeRequest](https://github.com/Microsoft/thrifty/blob/master/thrifty-runtime/src/main/java/com/microsoft/thrifty/service/ClientBase.java#L219-L270) reusable.
1) it uses package-private APIs of `MethodCall`
2) it's nicer to share that code instead of copying it

This is the modified `ClientBase` I've written (note: it's not used in production and I'm not sure how well tested it is):
```java
package com.microsoft.thrifty.service;

import com.microsoft.thrifty.ThriftException;
import com.microsoft.thrifty.protocol.MessageMetadata;
import com.microsoft.thrifty.protocol.Protocol;
import java.io.Closeable;
import java.io.IOException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public class BlockingClientBase implements Closeable {

    /**
     * A sequence ID generator; contains the most-recently-used
     * sequence ID (or zero, if no calls have been made).
     */
    private final AtomicInteger seqId = new AtomicInteger(0);

    /**
     * A flag indicating whether the client is active and connected.
     */
    private final AtomicBoolean running = new AtomicBoolean(true);

    private final Protocol protocol;

    protected BlockingClientBase(Protocol protocol) {
        this.protocol = protocol;
    }

    @Override
    public void close() {
        if (!running.compareAndSet(true, false)) {
            return;
        }

        try {
            protocol.close();
        } catch (IOException ignored) {
            // nope
        }
    }

    protected <T> Result<T> enqueue(MethodCall<T> methodCall) {
        if (!running.get()) {
            throw new IllegalStateException("Cannot write to a closed service client");
        }

        try {
            return invokeRequest(methodCall);
        } catch (IOException e) {
            close();
            throw new RuntimeException(e);
        }
    }

    private <T> Result<T> invokeRequest(MethodCall<T> call) throws IOException {
        boolean isOneWay = call.callTypeId == TMessageType.ONEWAY;
        int sid = seqId.incrementAndGet();

        protocol.writeMessageBegin(call.name, call.callTypeId, sid);
        call.send(protocol);
        protocol.writeMessageEnd();
        protocol.flush();

        if (isOneWay) {
            // No response will be received
            return Result.createWithResult(null);
        }

        MessageMetadata metadata = protocol.readMessageBegin();

        if (metadata.seqId != sid) {
            throw new ThriftException(
                    ThriftException.Kind.BAD_SEQUENCE_ID,
                    "Unrecognized sequence ID");
        }

        if (metadata.type == TMessageType.EXCEPTION) {
            ThriftException e = ThriftException.read(protocol);
            protocol.readMessageEnd();
            return Result.createWithError(e);
        } else if (metadata.type != TMessageType.REPLY) {
            throw new ThriftException(
                    ThriftException.Kind.INVALID_MESSAGE_TYPE,
                    "Invalid message type: " + metadata.type);
        }

        if (metadata.seqId != seqId.get()) {
            throw new ThriftException(
                    ThriftException.Kind.BAD_SEQUENCE_ID,
                    "Out-of-order response");
        }

        if (!metadata.name.equals(call.name)) {
            throw new ThriftException(
                    ThriftException.Kind.WRONG_METHOD_NAME,
                    "Unexpected method name in reply; expected " + call.name
                            + " but received " + metadata.name);
        }

        T result = null;
        Exception error = null;
        try {
            result = call.receive(protocol, metadata);
        } catch (Exception e) {
            error = e;
        }


        if (error != null) {
            return Result.createWithError(error);
        } else {
            return Result.createWithResult(result);
        }
    }

    protected static final ServiceMethodCallback<Object> STUB = new ServiceMethodCallback<Object>() {
        @Override
        public void onSuccess(Object result) {
            throw new AssertionError("STUB. Should never be called.");
        }

        @Override
        public void onError(Throwable error) {
            throw new AssertionError("STUB. Should never be called.");
        }
    };
}
```

**Comment by benjamin-bader** [comment]
**TL;DR:**
Thanks for contributing, very happy to have you here!  Synchronous services would be a valuable addition.  I don't think call adapters are the right approach; we're already generating very specific code, so we may as well teach the compiler to generate the runtime code we want.

I Have Thoughts about what this might look like, but welcome alternate ideas too.

I'm not able to work on code over the next few months, but will be thrilled to take PRs and am happy to offer guidance.

**The Long Version:**
Synchronous service calls sound like a great idea.  I'm not very excited to support `Future` in the interface; it seems best (IMHO) to generate interfaces that either return a result or throw.  I like the idea of supporting multiple interface styles, but am not sure how runtime call-adapters fit in.  Since we're generating the interfaces ourselves, why not also generate appropriate implementations?

This would be a bit of a job, since (as you've noted) the runtime currently assumes that all service calls are async.  @gabrielittner hit the nail on the head there - we'll need to refactor `ClientBase` into something more composable.  I'm not sure what the best approach there will be, but I'm confident that there won't be major problems there.

After letting this sit in the back of my mind for a few hours, here's one idea.  We could add a compiler option, e.g. `--service-style=[default|sync|rx|etc]`, and generate appropriate service code based on it.  `default` would map to the current implementation, `sync` to something synchronous, and the rest could follow as needed.

An example of the kind of interfaces I'm imagining:

```thrift
// IDL
service Foo {
  1: i32 doFoo(1: i64 arg) throws (1: BarException);
}
```

```java
// Async interface
interface Foo {
  void doFoo(long arg, ServiceMethodCallback<Integer> callback);
}
```

```java
// Synchronous interface
interface Foo {
  int doFoo() throws IOException, BarException;
}
```

Nothing too controversial there, I imagine.  The high-level work list for this would be:

1. Refactor `ClientBase` logic into something generally re-usable
2. Extract an interface out of `ServiceBuilder`, and make `DefaultServiceBuilder` and `SynchronousServiceBuilder` implementations
3. Add the appropriate compiler flags
4. Add any implementation-specific runtime support to a new library - e.g. `thrifty-rx` for any rx-specific bits.

The extra runtime libs would be a pain point, but not more than, say, `retrofit-gson`.  If we ever get around to #120, the pain would vanish.

All that said, I won't be able to do much work here in the next few months - non-work stuff is occupying my attention at least through August.  I'm happy to take PRs, and to offer what guidance I can.  I know these are big chunks of work, and particularly the codegen code can be daunting (in that it is overly complex and poorly documented - I'm definitely not thrilled with the implementation here).

Sorry for the long brain-dump - I didn't have enough time to write something more concise :)

**Comment by gotev** [comment]
@benjamin-bader 

> Synchronous service calls sound like a great idea. I'm not very excited to support Future in the interface; it seems best (IMHO) to generate interfaces that either return a result or throw.

I completely agree with that

> After letting this sit in the back of my mind for a few hours, here's one idea. We could add a compiler option, e.g. --service-style=[default|sync|rx|etc], and generate appropriate service code based on it. default would map to the current implementation, sync to something synchronous, and the rest could follow as needed.

This sounds better to me than the runtime call adapters. We can save methods count and reduce overhead and complexity of the generated code.

For the steps you mentioned above, I will need further guidance on how to proceed.

**Comment by benjamin-bader** [comment]
Cool!  I can't tell what guidance you need, so please feel free to ask detailed questions.  I'll expand a bit on what I think needs to be done, though I'm sure it won't answer everything:

So, the first thing to do is probably the most subtle, which is to figure out what bits of the service-client runtime are common and re-usable.  That means that `ClientBase` will probably be split into an `AsyncClientBase`, which will handle the work-queue aspects of the current `ClientBase`, and an actual base class that will provide support for the call-response protocol stuff.

`ServiceBuilder` is the class in the `thrifty-java-codegen` module that does the work of generating service implementations.  It will need to be abstracted a bit, so that `ThriftCodeGenerator` (the main codegen class) can just drop in whatever service-builder implementation (async, sync, rx, etc) is needed, based on its configuration.

The compiler flags are the easy part - `ThriftCompiler` is just a simple entry-point for driving `ThriftyCodeGenerator`; adding new flags is very simple.

For doing synchronous services, we won't need any additional runtime support - that would really only apply to rx implementations, I think.

**Comment by gabrielittner** [comment]
@gotev Did you already start with anything? Otherwise I'd take a take stab at the `ClientBase` changes.

**Comment by gotev** [comment]
@gabrielittner not yet, been really busy unfortunately. Hope to start in 1-2 weeks, so if you want to start, go ahead ;)

**Comment by benjamin-bader** [comment]
Closing this as "wontfix".
